### PR TITLE
Eliminate per-quote/per-trade allocations on the leaf hot path (#80)

### LINKED ISSUE #80: book: eliminate per-quote/per-trade allocations on the leaf hot path
## Overview

best_quote reads top-of-book size via bid_depth_at_price/ask_depth_at_price, each calling get_volume_by_price() which iterates the whole bid AND ask SkipMaps and allocates two HashMaps — so a single quote read does two full book scans and four HashMap allocations, plus Quote::new mints an unused ULID (clock+RNG) per call. The trade-capture listener is always installed and clones the full TradeResult under a Mutex on every match even on the non-_full hot path. Quote accessors lack #[inline] for cross-crate inlining, and the dead update_last_quote(&mut self) cache is unreachable through Arc. lib.rs/README also wrongly claim best_quote is O(1) with caching.

## Context

src/orderbook/book.rs:229-234/1235-1248/1351/1372-1382/1459, src/orderbook/quote.rs:55/73-118, src/lib.rs:205. Use total_depth_at_levels(1, side) or a single get_volume_by_price snapshot; mark quote reads #[inline]; remove the unused Quote id; gate trade capture behind a cheap armed flag or opt-in; resolve or remove the dead last_quote API; fix the O(1) doc claim.

## Tasks

- [ ] Compute best-bid/ask size with total_depth_at_levels(1, side) (or one get_volume_by_price snapshot reused for both)
- [ ] Remove the unused OrderId from Quote (clock+RNG cost; makes serialization deterministic)
- [ ] Add a has_both_sides() helper so the strike layer stops calling best_quote to test two-sidedness
- [ ] Mark best_quote/best_bid/best_ask/mid_price/spread and the Quote Copy accessors #[inline]
- [ ] Make trade capture opt-in or gate the clone+lock behind a cheap armed atomic
- [ ] Resolve the dead update_last_quote(&mut self) cache (interior mutability with &self, or remove it)
- [ ] Fix the lib.rs/README 'O(1) with caching' best_quote claim
- [ ] Add a one-sided-book best_quote test (only bid, only ask) asserting the empty side is None and is_two_sided() is false

## Acceptance criteria

- [ ] best_quote performs no full-book HashMap allocation and at most one ordered top-of-book read per side
- [ ] Criterion best_quote/update_last_quote cases show reduced allocations/time before vs after (make bench numbers attached)
- [ ] One-sided best_quote test passes
- [ ] cargo clippy --all-targets --all-features --workspace -- -D warnings clean
- [ ] cargo test --all-features passes
- [ ] make readme regenerated (O(1) claim fixed)

---

**Priority:** high · **Effort:** m · **Subsystem:** hierarchy · **Kind:** performance

_Filed from a multi-agent code-review audit of the codebase._

### PR DESCRIPTION
**Pull request by joaquinbejar**
## Summary

`best_quote` read top-of-book size via `bid_depth_at_price`/`ask_depth_at_price`,
each calling `get_volume_by_price()` which iterates the whole bid **and** ask
SkipMaps and allocates two HashMaps — so a single quote read did two full book
scans and four HashMap allocations, plus `Quote::new` minted an unused ULID
(clock + RNG) per call, and the trade-capture listener cloned the full
`TradeResult` under a `Mutex` on every match. This removes all of that from the
leaf hot path.

## Changes

- **`best_quote` allocation-free.** Each populated side's best-level size now
  comes from `total_depth_at_levels(1, side)` (one ordered top-of-book read, no
  HashMap; verified stack-iterator / no heap alloc in orderbook-rs 0.9.0); the
  empty side short-circuits to `(None, 0)`.
- **`Quote` drops its unused `OrderId`.** Kills the per-call ULID clock+RNG and
  makes `Quote` serialization deterministic.
- **Trade capture gated behind `Arc<AtomicBool>` (default disarmed).** The
  listener does one `Relaxed` load and returns before any clone/lock when
  disarmed. New API `arm_trade_capture` / `is_trade_capture_armed`; the `_full`
  methods arm via an RAII scope guard that restores the prior state on drop.
- **`has_both_sides()`** added on the leaf; the strike layer's `is_fully_quoted()`
  switched to it (no longer builds two `Quote`s to test two-sidedness).
- `#[inline]` on the quote accessors; **dead `update_last_quote`/`last_quote`
  cache removed**; the false **"O(1) with caching"** doc claim fixed.

## Performance (Criterion, `orderbook_operations`)

| case | before | after | change |
|------|--------|-------|--------|
| `best_quote_populated` (100×2 levels) | 11.751 µs | 69.10 ns | **~170× faster** |
| `best_quote_empty` | 67.09 ns | 37.51 ns | ~44% faster |
| `best_quote_one_sided` (new) | — | 56.49 ns | — |
| `update_last_quote` | 246.63 ns | removed (dead API) | — |

`best_quote` now does no full-book scan and no heap allocation, with at most one
ordered top-of-book read per populated side.

## Public API impact (breaking, within the 0.4.4 → 0.5.0 window)

- **Removed:** `Quote::id()` + the `Quote.id` field; `OptionOrderBook::update_last_quote`/`last_quote`/`last_quote_arc` (all dead — zero callers outside the leaf/tests/bench).
- **Added:** `OptionOrderBook::has_both_sides`, `arm_trade_capture`, `is_trade_capture_armed`.
- **Behavior change:** trade capture is **disarmed by default**, so `last_trade_result()` no longer auto-populates on the plain order path. Pre-0.5.0 pollers must call `arm_trade_capture(true)` or use the NATS trade feed / `_full` return values / order-state tracker (all independent of the flag). Documented on `last_trade_result`/`arm_trade_capture`.
- `Quote`'s serde shape loses `id` (deterministic) — `Quote` is **not** journaled or published, so no replay/wire impact. `src/lib.rs` doc + `README.md` regenerated (O(1) claim fixed). Version stays `0.5.0`.

## Testing

- [x] One-sided-book `best_quote` test (bid-only, ask-only, empty) asserting the empty side is `None`/`0` and `is_two_sided()`/`has_both_sides()` are false
- [x] Best-level-sum test (multiple orders at best level summed; deeper level excluded)
- [x] Trade-capture disarmed-by-default + `_full`-captures-regardless tests
- [x] Criterion before/after attached
- [x] `make pre-push` clean; `cargo build --features nats,sequencer` clean

## Checklist

- [x] Code follows `rules/global_rules.md` and `CLAUDE.md`
- [x] No `.unwrap()`/`.expect()`/unchecked indexing in production code
- [x] No `unsafe`; `tracing` only
- [x] Layering respected (`has_both_sides` on the leaf; strike calls down; leaf knows nothing of strike/chain)
- [x] Matching delegated to `orderbook-rs`
- [x] No new dependencies / feature flags
- [x] `src/lib.rs` docs + `README.md` (via `make readme`) updated

Closes #80

### CODE REVIEW

File: src/orderbook/book.rs [hunk 1]
```diff
@@ -993,6 +1074,21 @@ impl OptionOrderBook {
...
+    /// Used by the `_full` order methods so a trade is captured even when the
+    /// book is not under continuous capture. The guard restores (rather than
+    /// unconditionally disarming) so a `_full` call nested inside an armed scope
+    /// leaves capture armed afterwards.
+    fn arm_capture_scope(&self) -> CaptureArmGuard<'_> {
+        let previous = self.trade_capture_armed.swap(true, Ordering::Relaxed);
+        CaptureArmGuard {
+            flag: &self.trade_capture_armed,
+            previous,
+        }
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply src/orderbook/book.rs:1103]: Real bug — fixed in f5f31d5. Replaced the swap/restore-prior boolean with a refcount: added `trade_capture_scopes: AtomicUsize`, the capture listener now gates on `manual_armed || scopes > 0`, and each `_full` scope guard does `fetch_add(1)` on entry / `fetch_sub(1)` on drop. Overlapping `_full` scopes on the same book now compose correctly — capture stays active until the last scope exits, no scope can leave the book stuck-armed, and the manual `arm_trade_capture` flag is never clobbered. Added `test_capture_scopes_refcount_composes` covering nested scopes and the no-stuck-armed property.

File: src/orderbook/quote.rs [hunk 2]
```diff
@@ -3,15 +3,17 @@
...
 //! two-sided markets (bid and ask).
 
-use orderbook_rs::OrderId;
 use serde::{Deserialize, Serialize};
 
 /// Represents a two-sided quote (bid and ask).
 ///
 /// A quote captures the best bid and ask prices and sizes at a point in time.
 /// Prices are in smallest units (e.g., cents, satoshis) as `u64`.
 ///
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply src/orderbook/quote.rs:13]: Fixed in f5f31d5 — the doc now states prices are `u128` (`Option<u128>`, `None` for an empty side) and sizes are `u64`, matching the field types.
